n3fit/src/n3fit/hyper_optimization/rewards.py

```diff
@@ -14,26 +17,137 @@
     best_worst: 3.4200
     std: 0.8925
 
 """
 import numpy as np
+from validphys.pdfgrids import xplotting_grid, distance_grids
 
 
-def average(fold_losses, **kwargs):
-    """ Returns the average of fold losses """
+def average(fold_losses=None, **_kwargs):
+    """Returns the average of fold losses"""
     return np.average(fold_losses)
 
 
-def best_worst(fold_losses, **kwargs):
-    """ Returns the maximum loss of all k folds """
+def best_worst(fold_losses=None, **_kwargs):
+    """Returns the maximum loss of all k folds"""
     return np.max(fold_losses)
 
 
-def std(fold_losses, **kwargs):
-    """ Return the standard dev of the losses of the folds """
+def std(fold_losses=None, **_kwargs):
+    """Return the standard dev of the losses of the folds"""
     return np.std(fold_losses)
+
+
+def fit_distance(n3pdfs=None, **_kwargs):
+    """Loss function for hyperoptimization based on the distance of
+    the fits of all folds to the first fold
+    """
+    if n3pdfs is None:
+        raise ValueError("fit_distance needs n3pdf models to act upon")
+    xgrid = np.concatenate([np.logspace(-6, -1, 20), np.linspace(0.11, 0.9, 30)])
+    plotting_grids = [xplotting_grid(pdf, 1.6, xgrid) for pdf in n3pdfs]
+    distances = distance_grids(n3pdfs, plotting_grids, 0)
+    # The first distance will obviously be 0
+    # TODO: define this more sensibly, for now it is just a template
+    max_distance = 0
+    for distance in distances:
+        max_distance = max(max_distance, distance.grid_values.max())
+    return max_distance
+
+
+def _set_central_value(n3pdf, model):
+    """Given an N3PDF object and a MetaModel, set the PDF_0 layer
+    to be the central value of the n3pdf object"""
+    from n3fit.backends import operations as op
+
+    # Get the input x
+    for key, grid in model.x_in.items():
+        if key != "integration_grid":
+            input_x = grid.numpy()
+            break
+
+    # Compute the central value of the PDF
+    full_pdf = n3pdf(input_x)
+    cv_pdf = op.numpy_to_tensor(np.mean(full_pdf, axis=0, keepdims=True))
+
+    def central_value(x, training=None):  # pylint: disable=unused-argument
+        return cv_pdf
+
+    model.get_layer("PDF_0").call = central_value
+    # This model won't be trainable ever again
+    model.trainable = False
+    model.compile()
+
+
+def fit_future_tests(n3pdfs=None, experimental_models=None, **_kwargs):
+    """Use the future tests as a metric for hyperopt
+
+    NOTE: this function should only be called once at the end of
+    every hyperopt iteration, as it is destructive for the models
+    """
+    if n3pdfs is None:
+        raise ValueError("fit_future_test needs n3pdf models to act upon")
+    if experimental_models is None:
+        raise ValueError("fit_future_test needs experimental_models to compute chi2")
+
+    from n3fit.backends import MetaModel
+
+    compatibility_mode = False
+    try:
+        import tensorflow as tf
+        from n3fit.backends import set_eager
```

Comment:
pylint complains about importing outside top-level. Also in another function where you import backeds.operations.

Reply:
I understand why, but currently pylints use is limited due to the huge number of warnings, so if we want to improve this we should prevent adding more of them. To stop pylint from complaining you don't need to move this to top-level, instead you can add `# pylint: disable=import-outside-toplevel`, maybe even in the `.pylintrc` if you prefer.

Reply:
In this file pylint only gives 5  warnings. I don't really think that limits its use.